### PR TITLE
feat(openapi): adapt graviti OpenAPI about draft

### PR DESCRIPTION
**Pull request by willandfree**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2j40ydf OpenAPIv2 对接 —— Draft](https://app.clickup.com/t/2j40ydf)

File: graviti/openapi/draft.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python3
```
Comment (graczhual) [inline graviti/openapi/draft.py:1]: add these functions to `__init__`

**Review by graczhual — CHANGES_REQUESTED** [review]

File: graviti/openapi/draft.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        ...     "draft-2",
+        ...     "main",
+        ... )
+        {
+            "draftNumber": 2
+        }
+
+    """
+    url = urljoin(url, f"{URL_PATH_PREFIX}/datasets/{dataset_name}/drafts")
+    post_data = {"branchName": branch_name, "title": title}
```
Comment (linjiX) [inline graviti/openapi/draft.py:50]: should the naming style be snake_case?
branchName -> branch_name

File: graviti/openapi/draft.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+    if state:
+        params["state"] = state
+    if branch_name:
+        params["branch_name"] = branch_name
+
+    return open_api_do("GET", access_key, url, params=params).json()  # type: ignore[no-any-return]
+
+
+def get_draft(url: str, access_key: str, dataset_name: str, *, draft_number: int) -> Dict[str, Any]:
```
Comment (graczhual) [inline graviti/openapi/draft.py:122]: do not need to add * here

File: graviti/openapi/draft.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        ...     2,
+        ...     status="CLOSED"
+        ... )
+
+    """
+    url = urljoin(url, f"{URL_PATH_PREFIX}/datasets/{dataset_name}/drafts/{draft_number}")
+    patch_data: Dict[str, Any] = {"draftNumber": draft_number}
+
+    if state:
+        patch_data["status"] = state
```
Comment (graczhual) [inline graviti/openapi/draft.py:206]: state

File: graviti/openapi/draft.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        ... )
+
+        Close the draft:
+
+        >>> update_draft(
+        ...     "https://graviti.com/",
+        ...     "ACCESSKEY-********",
+        ...     "MNIST",
+        ...     2,
+        ...     status="CLOSED"
```
Comment (graczhual) [inline graviti/openapi/draft.py:198]: state

File: graviti/openapi/draft.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+    open_api_do("PATCH", access_key, url, json=patch_data)
+
+
+def delete_draft(
+    url: str,
+    access_key: str,
+    dataset_name: str,
+    draft_number: int,
+) -> None:
+    """Execute the OpenAPI `DELETE /v1/datasets/{dataset_name}/drafts/{draft_number}`.
```
Comment (graczhual) [inline graviti/openapi/draft.py:221]: Deleting drafts should not be supported. There's a mistake in the documentation.

**Review by graczhual — CHANGES_REQUESTED** [review]

File: graviti/openapi/draft.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+from graviti.openapi.request import URL_PATH_PREFIX, open_api_do
+
+
+def create_draft(
+    url: str,
+    access_key: str,
+    dataset_name: str,
+    title: str,
+    branch_name: str,
```
Comment (graczhual) [inline graviti/openapi/draft.py:19]: branch_name is not required.

**Review by graczhual — CHANGES_REQUESTED** [review]

File: graviti/openapi/draft.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+
+from graviti.openapi.request import URL_PATH_PREFIX, open_api_do
+
+
+def create_draft(
+    url: str,
+    access_key: str,
+    dataset_name: str,
+    title: str,
+    branch_name: Optional[str] = None,
```
Comment (graczhual) [inline graviti/openapi/draft.py:19]: use * before branch_name

File: graviti/openapi/draft.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        {
+            "draftNumber": 2
+        }
+
+    """
+    url = urljoin(url, f"{URL_PATH_PREFIX}/datasets/{dataset_name}/drafts")
+    post_data = {"branch_name": branch_name, "title": title}
+
+    if description:
+        post_data["description"] = description
```
Comment (graczhual) [inline graviti/openapi/draft.py:53]: if branch_name

**Review by graczhual — CHANGES_REQUESTED** [review]

File: graviti/openapi/draft.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+    Returns:
+        The response of OpenAPI.
+
+    Examples:
+        >>> create_draft(
+        ...     "https://graviti.com/",
+        ...     "ACCESSKEY-********",
+        ...     "MNIST",
+        ...     "draft-2",
+        ...     "main",
```
Comment (graczhual) [inline graviti/openapi/draft.py:42]: branch_name=

File: graviti/openapi/draft.py [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+            "draftNumber": 2
+        }
+
+    """
+    url = urljoin(url, f"{URL_PATH_PREFIX}/datasets/{dataset_name}/drafts")
+    post_data = {"title": title}
+
+    if description:
+        post_data["description"] = description
+
```
Comment (graczhual) [inline graviti/openapi/draft.py:54]: do not need blank line

File: graviti/openapi/draft.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    description: Optional[str] = None,
+) -> Dict[str, int]:
+    """Execute the OpenAPI `POST /v1/datasets/{dataset_name}/drafts`.
+
+    Arguments:
+        url: The URL of the graviti website.
+        access_key: User's access key.
+        dataset_name: Name of the dataset, unique for a user.
+        title: The draft title.
+        branch_name: The branch name.
```
Comment (graczhual) [inline graviti/openapi/draft.py:30]: Add an explanation of what None means. like the state in list drafts

File: graviti/openapi/draft.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+    limit: int = 128,
+) -> Dict[str, Any]:
+    """Execute the OpenAPI `GET /v1/datasets/{dataset_name}/drafts`.
+
+    Arguments:
+        url: The URL of the graviti website.
+        access_key: User's access key.
+        dataset_name: Name of the dataset, unique for a user.
+        state: The draft state which includes "OPEN", "CLOSED", "COMMITTED", "ALL" and None.
+            where None means listing open drafts.
```
Comment (graczhual) [inline graviti/openapi/draft.py:80]: do not need where

**Review by graczhual — CHANGES_REQUESTED** [review]

File: graviti/openapi/draft.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+) -> Dict[str, int]:
+    """Execute the OpenAPI `POST /v1/datasets/{dataset_name}/drafts`.
+
+    Arguments:
+        url: The URL of the graviti website.
+        access_key: User's access key.
+        dataset_name: Name of the dataset, unique for a user.
+        title: The draft title.
+        branch_name: The specified branch name.
+            None means the default branch of the dataset.
```
Comment (graczhual) [inline graviti/openapi/draft.py:31]: do not need another line

**Review by graczhual — CHANGES_REQUESTED** [review]

**Review by graczhual — APPROVED** [review]

**Review by linjiX — APPROVED** [review]

**Pull request merged**